src/main/java/gov/nasa/jpf/constraints/solvers/ConstraintSolverService.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2015 United States Government, as represented by the Administrator
+ *                of the National Aeronautics and Space Administration. All Rights Reserved.
+ *           2017-2021 The jConstraints Authors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gov.nasa.jpf.constraints.solvers;
+
+import gov.nasa.jpf.constraints.api.ConstraintSolver;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** ServiceLoader for instantiating constraint solvers */
+public final class ConstraintSolverService {
+
+  private final ServiceLoader<ConstraintSolverProvider> loader;
+  private final Properties config = new Properties();
+  private final Map<String, ConstraintSolverProvider> providers =
+      new HashMap<String, ConstraintSolverProvider>();
+  private static final Logger logger = Logger.getLogger("constraints");
+  public static ConstraintSolverService service = new ConstraintSolverService();
+
+  private ConstraintSolverService() {
+    loader = ServiceLoader.load(ConstraintSolverProvider.class);
+    discoverProviders();
+  }
+
+  /** @return a Set of names of all the found SolverProvider services */
+  public Set<String> getNames() {
+    return providers.keySet();
+  }
+
+  /**
+   * Used to create an instance of the solver with the name "name" and the config "config".
+   *
+   * @param name The name of the solver that should be created.
+   * @param config The config the solver will receive.
+   * @return An instance of the specified Solver
+   * @throws SolverNotFoundExcpetion
+   */
+  public ConstraintSolver createSolver(String name, Properties config)
+      throws SolverNotFoundExcpetion {
+    ConstraintSolver solver = null;
+    ConstraintSolverProvider s = providers.get(name);
+    if (s == null) {
+      throw new SolverNotFoundExcpetion("Specified solver could not be found");
+    } else {
+      solver = s.createSolver(config);
+      return solver;
+    }
+  }
+
+  /** Used to save the available SolverProvider services */
+  public void registerProvider(ConstraintSolverProvider provider) {
```

Comment:
Should this be `public`? Each consumer can now add their own providers after the fact and create a program-wide side effect